frontend/spec/features/coupon_code_spec.rb

```diff
@@ -32,50 +32,106 @@ def create_basic_coupon_promotion(code)
-      end
+        it "informs about an invalid coupon code" do
+          fill_in "order_coupon_code", with: "coupon_codes_rule_man"
+          click_button "Save and Continue"
+          expect(page).to have_content(Spree.t(:coupon_code_not_found))
+        end
 
-      context "with a promotion" do
-        it "applies a promotion to an order" do
+        it "can enter an invalid coupon code, then a real one" do
+          fill_in "order_coupon_code", with: "coupon_codes_rule_man"
+          click_button "Save and Continue"
+          expect(page).to have_content(Spree.t(:coupon_code_not_found))
           fill_in "order_coupon_code", with: "onetwo"
           click_button "Save and Continue"
           expect(page).to have_content("Promotion (Onetwo)   -$10.00")
         end
+
+        context "with a promotion" do
+          it "applies a promotion to an order" do
+            fill_in "order_coupon_code", with: "onetwo"
+            click_button "Save and Continue"
+            expect(page).to have_content("Promotion (Onetwo)   -$10.00")
+          end
+        end
+      end
+
+      context 'as logged user' do
+        let!(:user) do
+          create(:user, bill_address: create(:address), ship_address: create(:address))
+        end
+
+        let!(:order) do
+          order = Spree::Order.create!(
+            email: "spree@example.com",
+            store: Spree::Store.first || FactoryGirl.create(:store)
+          )
+
+          order.reload
+          order.user = user
+          order.recalculate
+          order
+        end
+
+        before do
+          allow_any_instance_of(Spree::CheckoutController).to receive_messages(current_order: order)
+          allow_any_instance_of(Spree::CheckoutController).to receive_messages(try_spree_current_user: user)
+          allow_any_instance_of(Spree::OrdersController).to receive_messages(try_spree_current_user: user)
+        end
+
+        context 'with saved credit card' do
+          let(:bogus) { create(:credit_card_payment_method) }
+          let!(:credit_card) do
+            create(:credit_card, user_id: user.id, payment_method: bogus, gateway_customer_profile_id: "BGS-WEFWF")
+          end
+
+          before do
+            user.wallet.add(credit_card)
+
+            visit spree.root_path
+            click_link "RoR Mug"
+            click_button "add-to-cart-button"
+            order.reload
+            # To Cart
+            click_button "Checkout"
+            # To shipping method screen, address is auto-populated
+            # with user's saved addresses
+            click_button "Save and Continue"
+            # To payment screen
+            click_button "Save and Continue"
+          end
+
+          it "shows wallet payments on coupon code errors" do
+            fill_in "order_coupon_code", with: "coupon_codes_rule_man"
+            click_button "Save and Continue"
+
+            expect(page).to have_content(Spree.t(:coupon_code_not_found))
```

Comment:
I think it's helpful to use the actual string here. I've seen more than a few translation bugs get caught when people were specifying the key wrong in both places.